src/rgw/rgw_cloud_oss.cc

```diff
@@ -0,0 +1,635 @@
+  static int get_oss_resp_value(const std::string& http_response, const std::string& parent, 
+                                const std::string& name, std::string& value);
+
+public:
+  int send_data(void *ptr, size_t len);
+  
+  RGWRESTOSSRequest(CephContext *_cct, const std::string& _url, param_vec_t *_headers,
+                                 param_vec_t *_params, RGWCloudInfo& _cloud_info) 
+                                   : RGWRESTSimpleRequest(_cct, _url, _headers, _params),
+                                     lock("RGWRESTStreamWriteOSSRequest"), 
+                                     cloud_info(_cloud_info), bl_len(0),total_send(0),pos(0),
+                                     cb(NULL),
+                                     http_manager(_cct)
+  { }
+  ~RGWRESTOSSRequest();
+  
+  RGWGetDataCB *get_out_cb() { return cb; }
+
+  int put_obj(const std::string& bucket, const string& obj, bufferlist* bl, uint64_t obj_size);
+  int rm_obj(const std::string& bucket, const string& obj);
+  int create_bucket(const std::string& bucket);
+
+  int init_upload_multipart(const std::string& bucket, const std::string& key, std::string& upload_id, int& retcode);
+  int upload_multipart(const std::string& bucket, const std::string& key, const std::string& uplaod_id, 
+                       uint64_t seq, bufferlist& bl, uint64_t obj_size, std::string& etag);
+  int finish_upload_multipart(const std::string& bucket, const std::string& key, const std::string& upload_id,
+                              std::map<uint64_t, std::string>& etags);
+  int abort_upload_multipart(const std::string& bucket, const std::string& key, const std::string& upload_id); 
+};
+
+RGWRESTOSSRequest::~RGWRESTOSSRequest() {
+  delete cb;
+  cb = nullptr; 
+}
+
+void RGWRESTOSSRequest::reset() {
+  pos = 0;
+  total_send = 0;
+  cur_it = end_it;
+}
+
+int RGWRESTOSSRequest::add_output_data(bufferlist& bl, uint64_t len) {
+  if (status < 0) {
+    int ret = status;
+    return ret;
+  }
+  cur_it = bl.buffers().begin();
+  end_it = bl.buffers().end();
+  bl_len = len;
+  bool done;
+  return http_manager.process_requests(false, &done);
+}
+
+int RGWRESTOSSRequest::send_data(void *ptr, size_t len) {
+  uint64_t sent = 0;
+  if (bl_len == 0 || status < 0) {
+    dout(20) << "RGWRESTStreamWriteRequest::send_data status=" <<status << " len:"<<bl_len<<dendl;
+    reset();
+    return status;
+  }
+  
+  uint64 remain_len = 0;
+  while(cur_it != end_it && len>0) {
+    remain_len = cur_it->length()-pos;
+    uint64_t send_len = min(len, (size_t)remain_len);
+    memcpy(ptr, cur_it->c_str() + pos, send_len);
+    ptr = (char*)ptr + send_len;
+    len -= send_len;
+    sent += send_len;
+    pos += send_len;
+    total_send += send_len;
+     
+    if ( pos == cur_it->length() ) {
+      ++cur_it;
+      pos = 0;
+    }
+  }
+  
+  return sent;
+}
+
+int RGWRESTOSSRequest::complete() {
+  int ret = http_manager.complete_requests();
+  if (ret < 0)
+    return ret;
+  return status;
```

Comment:
ret is result of program excuting, status is the result remote service return. so if ret is ok, the status may be not